docs/performance/inserts/architecture.txt

```diff
@@ -0,0 +1,263 @@
+
+Bulk inserts are typically done via :ref:`the CrateDB HTTP endpoint
+<bulk_switching>`, but they can also be done with :ref:`client batching
+<bulk_client>` or with SQL via :ref:`UNNEST <bulk_unnest>`. Consult the
+respective docs for details of each method.
+
+All three bulk insert methods are handled the same internally by CrateDB.
+
+.. _processing_phases:
+
+Processing Phases
+=================
+
+.. _parsing_analyzing:
+
+Parsing and Analyzing
+---------------------
+
+When a user issues an insert statement to the cluster then this request is sent
+to whichever individual node the client connected to. This node is then
+responsible for performing statement parsing, analyzing, and planning.
+
+The statement is parsed by the *SQL parser*. The *parsed statement* is then
+analyzed by the *statement analyzer*, which produces an *analyzed statement*.
+
+The statement analyzer performs the following checks:
+
+- Does the target table exist?
+
+- Do the specified table columns exist?
+
+   - If not, does the table have a `dynamic column policy`_?
+
+- Do the data types of the insert values match the data types of the existing
+  columns?
+
+- Are all ``PRIMARY KEY`` columns and ``NOT NULL`` column specified?
+
+- Are the values for ``NOT NULL`` columns non null?
+
+- Is the table partitioned?
+
+  - If so, does a `partition`_ exist for all insert values?
+
+    - If not, enrich the statement with a list of the partitions that need to
+      be created
+
+.. _planning:
+
+Planning
+--------
+
+The analyzed statement is then handed off to the *planner*, which plans out the
+execution of the insert. This process continues to take place on the node
+responsible for handling the client request.
+
+How the planner proceeds depends on the type of insert being prepared.
+
+Planning Single Inserts
+.......................
+
+The planner performs the following actions:
+
+1. If the partition doesn't exist, create a new partition
+
+2. Create a single insert request
+
+Planning Inserts With Multiple Value Expressions
+................................................
+
+For each row to be inserted, the planner performs the following actions:
+
+1. If the partition doesn't exist, create a new partition
+
+2. Calculate the shard ID (``ShardId``) using the `routing column`_
+
+3. Create a single insert request for the shard ID
+
+Planning Bulk Inserts
+.....................
+
+The planner performs the same steps taken for inserts with multiple value
+expressions, as listed under the previous header.
+
+Additionally, the planner prepares a bulk insert response object that will hold
+the results of every row we try to insert.
+
+.. _execution:
+
+Execution
+---------
+
+The planned inserts are then handed off to the *executor*, which executes them
+across the cluster.
+
+This process is managed by the node responsible for handling the client
+request, but the operations needed to complete the inserts can be handed off to
+any `data node`_ in the cluster.
```

Comment:
hm.. I think here it would be nice to have some kind of illustration. E.g

```

    +--------+                  +---------+
    | Client |----- insert -->  | Handler |
    +--------+     (values)     +---------+
                                  |
                         ---------+----
                       _/              \_
                      /                  \
                     /                    \
                    insert               insert
                 +---------+          +---------+
                 |  Node   |          |  Node   |
                 | Shard 1 |          | Shard 2 |
                 +---------+          +---------+


    +--------+                 +---------+
    | Client |----- insert --> | Handler |
    +--------+    (subquery    +---------+
                   from table)   |
                         +--------++
                       _/           \_
                      /               \
                     /                 \
                    select           select
                 +---------+       +---------+
                 |  Node   |       |  Node   |
                 | Shard 1 |       | Shard 2 |
                 +---------+       +---------+
                    |  \            /  |
                    |   \          /   |
                    |    \________/    |
                    |    /        \    |
                    |   /          \   |
                    |  /            \  |
                    insert           insert
                 +---------+       +---------+
                 |  Node   |       |  Node   |
                 | Shard 1 |       | Shard 2 |
                 +---------+       +---------+
```

Reply:
yeah that would be nice. I think inserts-by-query are in general missing at this document yet

Reply:
this is being addressed with the rewrite